// rollup-plugin-postcss-lit/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const pluginutils_1 = require("@rollup/pluginutils");
const transformAst = require("transform-ast");
const escape = (str) => str
    .replace(/`/g, '\\`')
    .replace(/\\(?!`)/g, '\\\\');
function postcssLit(options = {}) {
    const defaultOptions = {
        include: '**/*.{css,sss,pcss,styl,stylus,sass,scss,less}',
        exclude: null,
        importPackage: 'lit-element'
    };
    const opts = Object.assign(Object.assign({}, defaultOptions), options);
    const filter = pluginutils_1.createFilter(opts.include, opts.exclude);
    return {
        name: 'postcss-lit',
        enforce: 'post',
        transform(code, id) {
            if (!filter(id))
                return;
            const ast = this.parse(code, {});
            // export default const css;
            let defaultExportName;
            // export default '...';
            let isDeclarationLiteral = false;
            const magicString = transformAst(code, { ast: ast }, node => {
                if (node.type === 'ExportDefaultDeclaration') {
                    defaultExportName = node.declaration.name;
                    isDeclarationLiteral = node.declaration.type === 'Literal';
                }
            });
            if (!defaultExportName && !isDeclarationLiteral) {
                return;
            }
            magicString.walk(node => {
                if (defaultExportName && node.type === 'VariableDeclaration') {
                    const exportedVar = node.declarations.find(d => d.id.name === defaultExportName);
                    if (exportedVar) {
                        exportedVar.init.edit.update(`cssTag\`${escape(exportedVar.init.value)}\``);
                    }
                }
                if (isDeclarationLiteral && node.type === 'ExportDefaultDeclaration') {
                    node.declaration.edit.update(`cssTag\`${escape(node.declaration.value)}\``);
                }
            });
            magicString.prepend(`import {css as cssTag} from '${opts.importPackage}';\n`);
            return {
                code: magicString.toString(),
                map: magicString.generateMap({
                    hires: true,
                }),
            };
        },
    };
}
exports.default = postcssLit;
;
