// @web/test-runner-cli/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestRunnerCli = exports.defaultReporter = exports.readCliArgsConfig = exports.validateCoreConfig = exports.readConfig = exports.startTestRunner = void 0;
var startTestRunner_1 = require("./startTestRunner");
Object.defineProperty(exports, "startTestRunner", { enumerable: true, get: function () { return startTestRunner_1.startTestRunner; } });
var readConfig_1 = require("./config/readConfig");
Object.defineProperty(exports, "readConfig", { enumerable: true, get: function () { return readConfig_1.readConfig; } });
Object.defineProperty(exports, "validateCoreConfig", { enumerable: true, get: function () { return readConfig_1.validateCoreConfig; } });
var readCliArgsConfig_1 = require("./config/readCliArgsConfig");
Object.defineProperty(exports, "readCliArgsConfig", { enumerable: true, get: function () { return readCliArgsConfig_1.readCliArgsConfig; } });
var defaultReporter_1 = require("./reporter/defaultReporter");
Object.defineProperty(exports, "defaultReporter", { enumerable: true, get: function () { return defaultReporter_1.defaultReporter; } });
var TestRunnerCli_1 = require("./cli/TestRunnerCli");
Object.defineProperty(exports, "TestRunnerCli", { enumerable: true, get: function () { return TestRunnerCli_1.TestRunnerCli; } });
