// @web/test-runner-core/dist/server/TestRunnerServer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestRunnerServer = void 0;
const dev_server_core_1 = require("@web/dev-server-core");
const chokidar_1 = __importDefault(require("chokidar"));
const watchFilesMiddleware_1 = require("./middleware/watchFilesMiddleware");
const cacheMiddleware_1 = require("./middleware/cacheMiddleware");
const testRunnerApiMiddleware_1 = require("./middleware/testRunnerApiMiddleware");
const serveTestRunnerHtmlPlugin_1 = require("./plugins/serveTestRunnerHtmlPlugin");
const serveTestFrameworkPlugin_1 = require("./plugins/serveTestFrameworkPlugin");
const CACHED_PATTERNS = [
    'node_modules/@web/test-runner-',
    'node_modules/mocha/',
    'node_modules/chai/',
];
const isDefined = (_) => Boolean(_);
class TestRunnerServer {
    constructor(config, sessions, testFiles, runSessions) {
        this.fileWatcher = chokidar_1.default.watch([]);
        const { plugins = [], testFramework, rootDir } = config;
        const { testFrameworkImport, testFrameworkPlugin } = testFramework
            ? serveTestFrameworkPlugin_1.serveTestFrameworkPlugin(testFramework)
            : {};
        const serverConfig = {
            port: config.port,
            hostname: config.hostname,
            rootDir,
            mimeTypes: config.mimeTypes,
            middleware: [
                testRunnerApiMiddleware_1.testRunnerApiMiddleware(sessions, rootDir, config, plugins),
                watchFilesMiddleware_1.watchFilesMiddleware({ runSessions, sessions, rootDir, fileWatcher: this.fileWatcher }),
                cacheMiddleware_1.cacheMiddleware(CACHED_PATTERNS, config.watch),
                ...(config.middleware || []),
            ],
            plugins: [
                serveTestRunnerHtmlPlugin_1.serveTestRunnerHtmlPlugin(config, testFiles, testFrameworkImport),
                testFrameworkPlugin,
                ...(config.plugins || []),
            ].filter(isDefined),
        };
        this.devServer = new dev_server_core_1.DevServer(serverConfig, config.logger, this.fileWatcher);
    }
    async start() {
        await this.devServer.start();
    }
    async stop() {
        await this.devServer.stop();
    }
}
exports.TestRunnerServer = TestRunnerServer;
