// @mdjs/mdjs-preview/MdJsPreview.js
import { LitElement, html, css } from 'lit-element';

export class MdJsPreview extends LitElement {
  static get properties() {
    return {
      story: {
        type: Function,
      },
      code: {
        type: String,
      },
      showCode: {
        type: Boolean,
        attribute: 'show-code',
        reflect: true,
      },
    };
  }

  toggleShowCode() {
    this.showCode = !this.showCode;
  }

  constructor() {
    super();
    this.story = () => html` <p>Loading...</p> `;
  }

  render() {
    return html`
      <div id="wrapper">
        <div>${this.story()}</div>
        <button id="showCodeButton" @click=${this.toggleShowCode}>show code</button>
      </div>
      <pre id="codePre"><code>${this.code}</code></pre>
    `;
  }

  static get styles() {
    return css`
      :host {
        border: 1px solid #ccc;

        display: block;
        box-shadow: rgba(0, 0, 0, 0.1) 0px 1px 3px 0px;
      }

      #showCodeButton {
        position: absolute;
        right: 0;
        bottom: 0;
      }

      #wrapper {
        position: relative;
        padding: 10px;
      }

      #codePre {
        display: none;
        margin: 0;
        padding: 16px;
        overflow: auto;
        font-size: 85%;
        line-height: 1.45;
        background-color: #f6f8fa;
        border-radius: 3px;
      }

      :host([show-code]) #codePre {
        display: block;
      }
    `;
  }
}
