// @web/dev-server-cli/dist/config/readCliArgsConfig.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.readCliArgsConfig = void 0;
const command_line_args_1 = __importDefault(require("command-line-args"));
const command_line_usage_1 = __importDefault(require("command-line-usage"));
const camelcase_1 = __importDefault(require("camelcase"));
const defaultOptions = [
    {
        name: 'config',
        alias: 'c',
        type: String,
        description: 'The file to read configuration from. Config entries are camelCases flags.',
    },
    {
        name: 'root-dir',
        alias: 'r',
        type: String,
        description: 'The root directory to serve files from. Defaults to the current working directory.',
    },
    {
        name: 'open',
        alias: 'o',
        type: String,
        description: 'Opens the browser on app-index, root dir or a custom path.',
    },
    {
        name: 'app-index',
        alias: 'a',
        type: String,
        description: "The app's index.html file. When set, serves the index.html for non-file requests. Use this to enable SPA routing.",
    },
    {
        name: 'help',
        type: Boolean,
        description: 'Print help commands',
    },
];
function readCliArgsConfig(extraOptions = [], argv = process.argv) {
    const options = [...defaultOptions, ...extraOptions];
    const cliArgs = command_line_args_1.default(options, { argv });
    // when the open flag is used without arguments, it defaults to null. treat this as "true"
    if ('open' in cliArgs && typeof cliArgs.open !== 'string') {
        cliArgs.open = true;
    }
    if ('help' in cliArgs) {
        /* eslint-disable-next-line no-console */
        console.log(command_line_usage_1.default([
            {
                header: 'Web Dev Server',
                content: 'Dev Server for web development.',
            },
            {
                header: 'Usage',
                content: 'web-dev-server [options...]' + '\nwds [options...]',
            },
            { header: 'Options', optionList: options },
        ]));
        process.exit();
    }
    const cliArgsConfig = {};
    for (const [key, value] of Object.entries(cliArgs)) {
        cliArgsConfig[camelcase_1.default(key)] = value;
    }
    return cliArgsConfig;
}
exports.readCliArgsConfig = readCliArgsConfig;
